pennylane/compiler/compiler.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Compiler developer functions"""
+
+from typing import List
+from importlib import reload
+from collections import defaultdict
+import dataclasses
+import pkg_resources
+
+
+@dataclasses.dataclass
+class AvailableCompilers:
+    """This contains data of installed PennyLane compiler packages"""
+
+    names_entrypoints = {}
+
+
+def _refresh_compilers():
+    """Scan installed PennyLane compiler pacakges to refresh the compilers
+    names and entry points
+    """
+
+    reload(pkg_resources)
+
+    # Refresh the list of compilers
+    AvailableCompilers.names_entrypoints = defaultdict(dict)
+
+    # Iterator packages entry-points with the 'pennylane.compilers' group name
+    for entry in pkg_resources.iter_entry_points("pennylane.compilers"):
+        # Only need name of the parent module
+        module_name = entry.module_name.split(".")[0]
+        AvailableCompilers.names_entrypoints[module_name][entry.name] = entry
+
+
+def available_compilers() -> List[str]:
+    """Loads and returns a list of available compilers that are
+    installed and compatible with the :func:`~.qjit` decorator.
+
+    **Example**
+
+    This method returns the name of installed compiler packages supported in
+    PennyLane. For example, after installing the `Catalyst <https://github.com/pennylaneai/catalyst>`__
+    compiler, this will now appear as an available compiler:
+
+    >>> qml.compiler.available_compilers()
+    ['catalyst']
+    """
+
+    # This class updates the class variable names_entrypoints
+    _refresh_compilers()
```

Comment:
I'd argue that it should be called in this `available_compilers` function on every call to ensure that all available compilers are still available. This function doesn't call on every call in the `available` method as expected. 

> Once upon import should be sufficient

Sorry, I don't get it! importing what?

Reply:
> I'd argue that it should be called in this available_compilers function on every call to ensure that all available compilers are still available.

To be fair it is just a question of specification, how do we want this function to behave? It seems we don't really use it anywhere so far. If the intention is that:
- the user calls this function directly on occasion to see what compilers are available to them -> it's not a bad idea that it refreshes every time 👍 
- we call this function from other PL functions -> we need to be careful because that can easily lead to slowdowns

> Sorry, I don't get it! importing what?

Importing this module. PennyLane does the same thing here (this line executes exactly once, when PL is first imported): https://github.com/PennyLaneAI/pennylane/blob/4e1f7495853df620852a50ba2daa9bceeb4bd86c/pennylane/__init__.py#L163

Reply:
I'd say the former application is intended for this `available_compilers`. You can see that's something that the user occasionally use and it doesn't have any use cases in the implementation of the  compiler driver and the QJIT API.

> PennyLane does the same thing here (this line executes exactly once, when PL is first imported

Yes, and we do exactly the same in the performance critical methods: `available` and `active`.